source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,279 @@
+    }
+  }
+};
+
+// Requires all verifier
+class AllVerifierImpl : public BaseGroupVerifierImpl {
+public:
+  AllVerifierImpl(const JwtRequirementAndList& and_list, const AuthFactory& factory,
+                  const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                  const Extractor& extractor_for_allow_fail, const BaseVerifierImpl* parent)
+      : BaseGroupVerifierImpl(parent) {
+    for (const auto& it : and_list.requirements()) {
+      verifiers_.emplace_back(innerCreate(it, providers, factory, extractor_for_allow_fail, this));
+    }
+  }
+
+  void onComplete(const Status& status, ContextImpl& context) const override {
+    auto& completion_state = context.getCompletionState(this);
+    if (completion_state.is_completed_) {
+      return;
+    }
+    if (++completion_state.number_completed_children_ == verifiers_.size() ||
+        Status::Ok != status) {
+      completion_state.is_completed_ = true;
+      completeWithStatus(status, context);
+    }
+  }
+};
+
+// Match all, for requirement not set
+class AllowAllVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowAllVerifierImpl(const BaseVerifierImpl* parent) : BaseVerifierImpl(parent) {}
+
+  void verify(ContextSharedPtr context) const override {
+    onComplete(Status::Ok, static_cast<ContextImpl&>(*context));
+  }
+
+  void onComplete(const Status& status, ContextImpl& context) const override {
+    completeWithStatus(status, context);
+  }
+};
+
+VerifierPtr innerCreate(const JwtRequirement& requirement,
+                        const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                        const AuthFactory& factory, const Extractor& extractor_for_allow_fail,
+                        const BaseVerifierImpl* parent) {
+  std::string provider_name;
+  std::vector<std::string> audiences;
+  switch (requirement.requires_type_case()) {
+  case JwtRequirement::RequiresTypeCase::kProviderName:
+    provider_name = requirement.provider_name();
+    break;
+  case JwtRequirement::RequiresTypeCase::kProviderAndAudiences:
+    for (const auto& it : requirement.provider_and_audiences().audiences()) {
+      audiences.emplace_back(it);
+    }
+    provider_name = requirement.provider_and_audiences().provider_name();
+    break;
+  case JwtRequirement::RequiresTypeCase::kRequiresAny:
+    return std::make_unique<AnyVerifierImpl>(requirement.requires_any(), factory, providers,
+                                             extractor_for_allow_fail, parent);
+  case JwtRequirement::RequiresTypeCase::kRequiresAll:
+    return std::make_unique<AllVerifierImpl>(requirement.requires_all(), factory, providers,
+                                             extractor_for_allow_fail, parent);
+  case JwtRequirement::RequiresTypeCase::kAllowMissingOrFailed:
+    return std::make_unique<AllowFailedVerifierImpl>(
+        factory, extractor_for_allow_fail, requirement.allow_missing_or_failed().value(), parent);
```

Comment:
Just wonder, is `requirement.allow_missing_or_failed().value() == false` normal? I mean, if config use this oneof, should it always set it to `true`? What the behavior we expected if it is `false`?

Reply:
Good point.  It doesn't make sense to have allow_mising_and_fail to false.  How about we change this field to google.protobuf.Empty instead of bool.

Reply:
changed config to google.protobuf.Empty.